src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+                AccountCredentialWithoutProofs::Normal { cdv, .. } => &cdv.cred_key_info.keys,
+            };
+
+            let on_chain_public_key =
+                on_chain_keys
+                    .get(&key_index)
+                    .ok_or(SignatureError::MissingIndicesOnChain {
+                        credential_index: credential_index.index,
+                        key_index:        key_index.0,
+                    })?;
+
+            let VerifyKey::Ed25519VerifyKey(public_key) = *on_chain_public_key;
+
+            // Check that the public key in the `account_keys` map matches the public key on
+            // chain.
+            if signing_key.public() != public_key {
+                return Err(SignatureError::MismatchPublicKeyOnChain {
+                    credential_index:    credential_index.index,
+                    key_index:           key_index.0,
+                    expected_public_key: Box::new(public_key),
+                    actual_public_key:   Box::new(signing_key.public()),
+                });
+            };
+
+            // Generate the signature.
+            let signature = signing_key.sign(&message_hash);
+
+            // Check that the signature is valid to ensure
+            // that the public and private keys in the `account_keys` map match.
+            if !VerifyKey::from(public_key).verify(message_hash, &Signature::from(signature)) {
+                return Err(SignatureError::MismatchPublicPrivateKeys {
+                    credential_index: credential_index.index,
+                    key_index:        key_index.0,
+                });
+            }
+
+            // Insert the generated signature into the return map.
+            account_signatures
+                .sigs
+                .entry(credential_index.index)
+                .or_insert_with(|| CredentialSignatures {
+                    sigs: BTreeMap::new(),
+                })
+                .sigs
+                .insert(key_index.0, signature.into());
+        }
+    }
+
+    Ok(account_signatures)
+}
+
+/// Wrapper around the [`sign_as_account`] function to simplify the signature
+/// generation for single-signer accounts.
+pub async fn sign_as_single_signer_account(
+    client: v2::Client,
+    signer: AccountAddress,
+    signing_key: SigningKey,
+    message: &Message<'_>,
+    bi: BlockIdentifier,
+) -> Result<Signature, SignatureError> {
+    let keypair: KeyPair = KeyPair::from(signing_key);
+    // Generate account keys that have one keypair at index 0 in both maps.
+    let keypairs = AccountKeys::from(InitialAccountData {
+        keys:      [(KeyIndex(0), keypair)].into_iter().collect(),
+        threshold: SignatureThreshold::ONE,
+    });
+    let signature = sign_as_account(client, signer, keypairs, message, bi).await?;
+    // Accessing the maps at indices 0 is safe because we generated an
+    // AccountSignature with a keypair at index 0 at both maps.
+    Ok(signature.sigs[&0].sigs[&0].clone())
```

Comment:
I think you could use `.remove(&0).unwrap()` instead of indexing to avoid the clone.